### DISCUSSION TITLE
LGPLv3 and commercial Node.js software

### DISCUSSION THREAD

**Discussion by kpollich** [openpgpjs/openpgpjs#1532] [General]
Hi there! I'm working on some key/signature verification in [Kibana](https://github.com/elastic/kibana) and was hoping to use `openpgpjs` as we're already generating the signatures in question elsewhere using https://github.com/ProtonMail/gopenpgp. The problem, however, is that `openpgpjs` is [licensed](https://github.com/openpgpjs/openpgpjs/blob/main/LICENSE) under LGPLv3, and I don't think we can suitably adhere to the terms of this license in a Node.js environment. 

For example, we can't allow users to easily provide a modified or "interface compatible" version of `openpgpjs` in a Node environment, because we can't load it as a separate script file like we might be able to in a browser environment. See https://github.com/openpgpjs/openpgpjs/discussions/1427 for relevant discussion around Webpack/client-side JS.

Digging through issues, I see https://github.com/openpgpjs/openpgpjs/issues/10 from 2011. I'm curious if the thoughts around licensing have potentially shifted here, and if it'd be possible to discuss dual licensing the project under a more permissive license like MIT. If not, I'd be eager to hear if there is a recommended alternative implementation of PGP in Node.

A few related discussions
- https://github.com/openpgpjs/openpgpjs/discussions/1208
- https://github.com/openpgpjs/openpgpjs/discussions/1427
- https://github.com/openpgpjs/openpgpjs/discussions/1211

Thanks much!

**Comment by twiss** [comment]
Hey :wave: As a disclaimer, I'm not a lawyer, but:

> For example, we can't allow users to easily provide a modified or "interface compatible" version of openpgpjs in a Node environment, because we can't load it as a separate script file like we might be able to in a browser environment. See https://github.com/openpgpjs/openpgpjs/discussions/1427 for relevant discussion around Webpack/client-side JS.

I'm not sure I understand why not - `require()` on Node.js is quite similar to `await import()` on the web, and can load a separate script file. It's even possible to specify "peer dependencies" in package.json, which can be an entirely separate package which has to be installed separately (and can be optional or required). I'm not sure if that's necessary to satisfy the requirements of LGPL, but I assume it should be sufficient, as long as the user can replace OpenPGP.js with a fork, for example (which I think should be quite easy in that case — but maybe it's also possible with a normal dependency, not sure).

> Digging through issues, I see https://github.com/openpgpjs/openpgpjs/issues/10 from 2011. I'm curious if the thoughts around licensing have potentially shifted here, and if it'd be possible to discuss dual licensing the project under a more permissive license like MIT.

We considered it, and I wouldn't be opposed to it, but this would essentially require us to receive permission to do so from all copyright holders, i.e. all authors of OpenPGP.js and the code it was based on, such as the gpg4browsers code mentioned in the issue you linked. So I think it's probably not feasible.

> If not, I'd be eager to hear if there is a recommended alternative implementation of PGP in Node.

Not specifically; I would recommend interfacing with an implementation in another language, in that case. For example, you could use gopenpgp via the command line interface [gosop](https://github.com/ProtonMail/gosop).

**Reply by kpollich** [reply]
Thanks for the reply, @twiss. I think we're in agreement here after having considered the mechanism by which a consumer of Kibana might swap out its dependency on `openpgp` with a fork. Since we ship the `node_modules` folder with Kibana when you download an open-source distribution of it, it's relatively simple to completely replace the actual source directory of `openpgp` and have Kibana load your own fork at runtime. It's also relatively easy to simply update `package.json` and rebuild from source, but that can be a somewhat expensive operation. 

I appreciate your input, and I think we're on our way to greenlighting usage of `openpgp` based on your comments and further investigation around how a user might swap in their own fork. 

Thank you very much!
